src/v/security/gssapi_principal_mapper.cc

```diff
@@ -0,0 +1,505 @@
+              index_replace.begin(), index_replace.end(), index);
+            if (conv_result.ec != std::errc()) {
+                vlog(
+                  seclog.warn,
+                  "Bad format in username mapping in {}",
+                  std::string_view{index_replace.data(), index_replace.size()});
+                return std::nullopt;
+            } else if (index >= params.size()) {
+                vlog(
+                  seclog.warn,
+                  "Invalid index provided ({}).  Outside range of "
+                  "parameters (length {})",
+                  index,
+                  params.size());
+                return std::nullopt;
+            } else {
+                const auto& param_index = params[index];
+                result.append(param_index.data(), param_index.size());
+            }
+        }
+
+        // format_piece will be empty if we have parsed all of format_piece
+        // text_replace and index_replace will be empty if nothing was grouped
+        // into those but re2::RE2::Consume still returns true (and doesn't
+        // update format_piece)
+        if (
+          format_piece.empty()
+          || (text_replace.empty() && index_replace.empty())) {
+            break;
+        }
+    }
+
+    return result;
+}
+
+ss::sstring gssapi_rule::replace_substitution(
+  std::string_view base,
+  const std::regex& from,
+  std::string_view to,
+  repeat repeat_) {
+    ss::sstring replace_value;
+
+    std::regex_constants::match_flag_type const flags
+      = std::regex_constants::format_default
+        | (repeat_ ? std::regex_constants::match_default : std::regex_constants::format_first_only);
+
+    replace_value = std::regex_replace(base.data(), from, to.data(), flags);
+
+    return replace_value;
+}
+
+gssapi_principal_mapper::gssapi_principal_mapper(
+  std::string_view default_realm,
+  config::binding<std::optional<std::vector<ss::sstring>>>
+    principal_to_local_rules_cb)
+  : _default_realm(default_realm)
+  , _principal_to_local_rules_binding(std::move(principal_to_local_rules_cb))
+  , _rules{detail::parse_rules(
+      _default_realm, _principal_to_local_rules_binding())} {
+    _principal_to_local_rules_binding.watch([this]() {
+        _rules = detail::parse_rules(
+          _default_realm, _principal_to_local_rules_binding());
+    });
+}
+
+std::optional<ss::sstring>
+gssapi_principal_mapper::apply(gssapi_name name) const {
+    std::vector<std::string_view> params;
+    if (name.host_name().empty()) {
+        if (name.realm().empty()) {
+            return name.primary();
+        }
+        params = {name.realm(), name.primary()};
+    } else {
+        params = {name.realm(), name.primary(), name.host_name()};
+    }
+
+    for (const auto& r : _rules) {
+        if (auto result = r.apply(params); result.has_value()) {
+            return result;
+        }
+    }
+
+    vlog(seclog.warn, "No rules apply to {}, rules: {}", name, _rules);
```

Comment:
A kerberos principal used to authenticate did not match any rule -> User should see this and update rules

Reply:
The downside to logging like this is that it lacks context. For example, what connection did this occur on? One way to solve that is to return an appropriate error or throw an exception. Then in the kafka connection context we can log in a way that annotates the log message with connection context information.

If a principal doesn't match a rule, then having that connection information could be very useful: a system with 1000s of connection might be hard to figure out where the rogue principal is coming from. With connection information we can provide things like the kafka client ID and IP address of the offending user.

No need to change this right now in this PR, but something to consider in follow-ups.